### PR TITLE
Improved removed waypoint error message

### PR DESCRIPTION
**Pull request by Dadopur**
When removing a waypoint lines are allowed to cross.
Crossing lines turn red.
Red lines are removed when they no longer cross.
Confirm area button disabled when red lines exist.
Warning message is displayed when lines crossing.
Fixed bug where warning message pops incorrectly when redefining area.

### CODE REVIEW

File: src/front-end/react_base/src/JS/Helpers/maphelper.js [hunk 1]
```diff
@@ -40,8 +40,251 @@ export function boundsToView(bounds) {
...
+
+    // Do the check for every line on the map.
+    for (let i = 0; i < waypoints.length - 2; i++) {
+      if (vectorHelper(a, b, c, d, waypoints, i)) {
+        newCrossingLine = [waypoints[0], waypoints[index - 1]];
+        return newCrossingLine;
+      }
+    }
+
+    // Case 3: index is first element in waypoints.
```
Comment (albarv340) [inline src/front-end/react_base/src/JS/Helpers/maphelper.js:144]: Is this also first element? Or is it a middle element

File: src/front-end/react_base/src/JS/Helpers/maphelper.js [hunk 2]
```diff
@@ -40,8 +40,251 @@ export function boundsToView(bounds) {
...
+  return newCrossingLine;
+}
+
+/**
+ * Checks if red line (point1 -> point2) are still crossing after removing a waypoint.
+ *
+ * @param {*} point1 starting point of red line.
+ * @param {*} point2 end point of red line.
+ * @param {*} waypoints list of all waypoints on the map.
+ * @param {*} index index of waypoint about to be removed.
```
Comment (albarv340) [inline src/front-end/react_base/src/JS/Helpers/maphelper.js:191]: Some type inference could be helpful (change the {*} to {Array} for example) and "index index" could be replaced by "index"

File: src/front-end/react_base/src/JS/IMMMap.js [hunk 3]
```diff
@@ -247,19 +101,59 @@ class IMMMap extends React.Component {
...
+        !(line === tempRemovedLine1 || line === tempRemovedLine2) &&
+        (line[0] === firstWaypoint || line[1] === lastWaypoint) &&
+        (line[0] === lastWaypoint || [line[1] === firstWaypoint])
+      ) {
+        redLinesToBeRemoved.push(index);
+      }
+    });
+
+    // Remove red lines if placing a new waypoint removes intersection
+    for (const i of redLinesToBeRemoved.reverse()) {
```
Comment (albarv340) [inline src/front-end/react_base/src/JS/IMMMap.js:151]: Is reverse() needed?

File: src/front-end/react_base/src/JS/IMMMap.js [hunk 4]
```diff
@@ -274,19 +168,77 @@ class IMMMap extends React.Component {
...
+          this.props.store.areaWaypoints,
+          i
+        )
+      ) {
+        redLinesToBeRemoved.push(index);
       }
+    });
+
+    // Remove red lines if the one of its waypoints gets removed.
+    for (const i of redLinesToBeRemoved.reverse()) {
```
Comment (albarv340) [inline src/front-end/react_base/src/JS/IMMMap.js:214]: is this reverse() needed?

**Review by albarv340 — COMMENTED** [review]
Some initial comments, more may come after more thorough testing

**Comment by Edvin-Dyremark** [comment]
It is now allowed to place waypoints that result in intersecting lines. This improved user experience when trying to define areas more complex than simple circles or squares. All intersecting line are painted red instead of just one.

File: src/front-end/react_base/src/JS/Helpers/maphelper.js [hunk 5]
```diff
@@ -40,8 +40,104 @@ export function boundsToView(bounds) {
...
+  det = (c - a) * (s - q) - (r - p) * (d - b);
+  if (det === 0) {
+    return false;
+  }
+
+  // length_1 & length_2 = lengths to intersecting point of vectors.
+  length_1 = ((s - q) * (r - a) + (p - r) * (s - b)) / det;
+  length_2 = ((b - d) * (r - a) + (c - a) * (s - b)) / det;
+
+  // If intersecting point is farther away than original vectors' length, then lengths will not be between 0 and 1.
```
Comment (albarv340) [inline src/front-end/react_base/src/JS/Helpers/maphelper.js:134]: How does that relate to if vectors are crossing?

File: src/front-end/react_base/src/JS/IMMMap.js [hunk 6]
```diff
@@ -215,6 +63,28 @@ class IMMMap extends React.Component {
...
+        getDrones((response) => {
+          this.setState({ oldDrones: this.state.drones });
+          this.setState({ drones: response.arg.drones });
+        });
+      }, updateDronesTimer),
+    });
+  }
+
+  /**
+   * Remove double timer from componentDidMount
```
Comment (albarv340) [inline src/front-end/react_base/src/JS/IMMMap.js:82]: Clears timer when component is unmounted

**Review by albarv340 — APPROVED** [review]

**Pull request merged**